// jest-editor-support/build/parsers/helper.js
"use strict";

require("core-js/modules/es.array.is-array.js");
require("core-js/modules/es.symbol.js");
require("core-js/modules/es.symbol.description.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.symbol.iterator.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.string.iterator.js");
require("core-js/modules/web.dom-collections.iterator.js");
require("core-js/modules/es.array.from.js");
require("core-js/modules/es.array.slice.js");
require("core-js/modules/es.date.to-string.js");
require("core-js/modules/es.regexp.to-string.js");
require("core-js/modules/es.function.name.js");
require("core-js/modules/es.object.define-property.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tsxPlugins = exports.tsPlugins = exports.parseOptions = exports.jsPlugins = exports.DefaultDecoratorPlugin = void 0;
require("core-js/modules/es.array.concat.js");
require("core-js/modules/es.regexp.exec.js");
require("core-js/modules/es.string.match.js");
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

// taken from https://github.com/babel/babel/blob/main/packages/babel-parser/typings/babel-parser.d.ts
// but comment out a few file-based and either-or plugins
var commonPlugins = ['asyncDoExpressions', 'asyncGenerators', 'bigInt', 'classPrivateMethods', 'classPrivateProperties', 'classProperties', 'classStaticBlock',
// Enabled by default
'decimal',
// 'decorators-legacy',
'decoratorAutoAccessors', 'destructuringPrivate', 'doExpressions', 'dynamicImport', 'explicitResourceManagement', 'exportDefaultFrom', 'exportNamespaceFrom',
// deprecated
// 'flow',
'flowComments', 'functionBind', 'functionSent', 'importMeta',
// 'jsx',
'logicalAssignment', 'importAssertions', 'importReflection', 'moduleBlocks', 'moduleStringNames', 'nullishCoalescingOperator', 'numericSeparator', 'objectRestSpread', 'optionalCatchBinding', 'optionalChaining', 'partialApplication',
// 'placeholders',
'privateIn',
// Enabled by default
'regexpUnicodeSets', 'throwExpressions', 'topLevelAwait', 'v8intrinsic',
// plugin with options
// ['decorators', {decoratorsBeforeExport: true}],
['pipelineOperator', {
  proposal: 'smart'
}], 'recordAndTuple'];
var DefaultDecoratorPlugin = exports.DefaultDecoratorPlugin = ['decorators', {
  decoratorsBeforeExport: true
}];
var jsPlugins = exports.jsPlugins = [].concat(commonPlugins, ['flow', 'jsx']);
var tsPlugins = exports.tsPlugins = [].concat(commonPlugins, ['typescript']);
var tsxPlugins = exports.tsxPlugins = [].concat(commonPlugins, ['typescript', 'jsx']);
var parseOptions = exports.parseOptions = function parseOptions(filePath, options) {
  var optionalPlugins = function optionalPlugins() {
    var _options$plugins, _options$plugins2;
    if (!(options !== null && options !== void 0 && (_options$plugins = options.plugins) !== null && _options$plugins !== void 0 && _options$plugins.decorators)) {
      return [DefaultDecoratorPlugin];
    }
    if (((_options$plugins2 = options.plugins) === null || _options$plugins2 === void 0 ? void 0 : _options$plugins2.decorators) === 'legacy') {
      return ['decorators-legacy'];
    }
    return [['decorators', options.plugins.decorators]];
  };
  if (filePath.match(/\.ts$/i)) {
    return {
      plugins: [].concat(_toConsumableArray(tsPlugins), _toConsumableArray(optionalPlugins()))
    };
  }
  if (filePath.match(/\.tsx$/i)) {
    return {
      plugins: [].concat(_toConsumableArray(tsxPlugins), _toConsumableArray(optionalPlugins()))
    };
  }

  // for backward compatibility, use js parser as default unless in strict mode
  if (!(options !== null && options !== void 0 && options.strictMode) || filePath.match(/\.m?jsx?$/i)) {
    return {
      plugins: [].concat(_toConsumableArray(jsPlugins), _toConsumableArray(optionalPlugins()))
    };
  }
  throw new TypeError("unable to find parser options for unrecognized file extension: ".concat(filePath));
};
