// jest-editor-support/build/Process.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
require("core-js/modules/es.symbol.js");
require("core-js/modules/es.symbol.description.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.symbol.iterator.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.string.iterator.js");
require("core-js/modules/web.dom-collections.iterator.js");
require("core-js/modules/es.array.from.js");
require("core-js/modules/es.array.slice.js");
require("core-js/modules/es.date.to-string.js");
require("core-js/modules/es.regexp.to-string.js");
require("core-js/modules/es.function.name.js");
require("core-js/modules/es.regexp.exec.js");
require("core-js/modules/es.symbol.to-primitive.js");
require("core-js/modules/es.date.to-primitive.js");
require("core-js/modules/es.number.constructor.js");
require("core-js/modules/es.object.define-property.js");
require("core-js/modules/es.object.keys.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.object.get-own-property-descriptor.js");
require("core-js/modules/es.array.for-each.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.define-properties.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createProcess = void 0;
require("core-js/modules/es.array.is-array.js");
require("core-js/modules/es.array.concat.js");
require("core-js/modules/es.array.join.js");
var _child_process = require("child_process");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; } /**
                                                                                                                                                                                     * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
                                                                                                                                                                                     *
                                                                                                                                                                                     * This source code is licensed under the MIT license found in the
                                                                                                                                                                                     * LICENSE file in the root directory of this source tree.
                                                                                                                                                                                     *
                                                                                                                                                                                     */
// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
var isLoginShell = function isLoginShell(arg) {
  return arg && typeof arg.path === 'string' && Array.isArray(arg.args);
};

/**
 * Spawns and returns a Jest process with specific args
 *
 * @param {string[]} args
 * @returns {ChildProcess}
 */
// eslint-disable-next-line import/prefer-default-export
var createProcess = exports.createProcess = function createProcess(workspace, args) {
  var _workspace$nodeEnv;
  var runtimeExecutable = [workspace.jestCommandLine].concat(_toConsumableArray(args));

  // If a path to configuration file was defined, push it to runtimeArgs
  if (workspace.pathToConfig) {
    runtimeExecutable.push('--config');
    runtimeExecutable.push(workspace.pathToConfig);
  }
  var env = _objectSpread(_objectSpread({}, process.env), (_workspace$nodeEnv = workspace.nodeEnv) !== null && _workspace$nodeEnv !== void 0 ? _workspace$nodeEnv : {});
  var cmd = runtimeExecutable.join(' ');
  var spawnCommandLine = function spawnCommandLine() {
    var spawnOptions = {
      cwd: workspace.rootPath,
      env: env,
      shell: typeof workspace.shell === 'string' && workspace.shell ? workspace.shell : true,
      // for non-windows: run in detached mode so the process will be the group leader and any subsequent process spawned
      // within can be later killed as a group to prevent orphan processes.
      // see https://nodejs.org/api/child_process.html#child_process_options_detached
      detached: process.platform !== 'win32'
    };
    if (workspace.debug) {
      // eslint-disable-next-line no-console
      console.log("spawning process with command=".concat(cmd), 'options:', spawnOptions);
    }
    return (0, _child_process.spawn)(cmd, [], spawnOptions);
  };
  var spawnLoginShell = function spawnLoginShell(shell) {
    var spawnOptions = {
      cwd: workspace.rootPath,
      env: env,
      detached: process.platform !== 'win32'
    };
    if (workspace.debug) {
      // eslint-disable-next-line no-console
      console.log("spawning login-shell \"".concat(shell.path, " ").concat(shell.args.join(' '), "\" for command=").concat(cmd), 'options:', spawnOptions);
    }
    var child = (0, _child_process.spawn)(shell.path, shell.args, spawnOptions);
    child.stdin.write("".concat(cmd, " \nexit $?\n"));
    return child;
  };
  if (isLoginShell(workspace.shell)) {
    if (process.platform === 'win32') {
      console.error('currently login-shell is only supported for non-windown platforms');
    } else {
      return spawnLoginShell(workspace.shell);
    }
  }
  return spawnCommandLine();
};
